src/Pact/Native/Ops.hs

```diff
@@ -34,30 +34,97 @@ import Pact.Types.Runtime
      \`(+ 1 2)` `(+ 5.0 0.5)` `(+ \"every\" \"body\")` `(+ [1 2] [3 4])` \
      \`(+ { \"foo\": 100 } { \"foo\": 1, \"bar\": 2 })`"
+  where
+
+    plus :: RNativeFun e
+    plus _ [TLitString a,TLitString b] = return (tStr $ a <> b)
+    plus _ [TList a aty _,TList b bty _] = return $ TList
+      (a <> b)
+      (if aty == bty then aty else TyAny)
+      def
+    plus _ [TObject as aty _,TObject bs bty _] =
+      let reps (a,b) = (abbrev a,(a,b))
+          mapit = M.fromList . map reps
+      in return $ TObject
+           (M.elems $ M.union (mapit as) (mapit bs))
+           (if aty == bty then aty else TyAny)
+           def
+    plus i as = binop' (+) (+) i as
+    {-# INLINE plus #-}
+
+    plusTy :: FunTypes n
+    plusTy = coerceBinNum <> binTy plusA plusA plusA
+
+subDef :: NativeDef
 subDef = defRNative "-" minus (coerceBinNum <> unaryNumTys)
      "Negate X, or subtract Y from X. `(- 1.0)` `(- 3 2)`"
+  where
+
+    minus :: RNativeFun e
+    minus _ [TLiteral (LInteger n) _] = return (toTerm (negate n))
+    minus _ [TLiteral (LDecimal n) _] = return (toTerm (negate n))
+    minus i as = binop' (-) (-) i as
+    {-# INLINE minus #-}
+
+    unaryNumTys :: FunTypes n
+    unaryNumTys = unaryTy numA numA
+
+mulDef :: NativeDef
 mulDef = defRNative "*" (binop' (*) (*)) coerceBinNum
      "Multiply X by Y. `(* 0.5 10.0)` `(* 3 5)`"
+
+divDef :: NativeDef
 divDef = defRNative "/" divide' coerceBinNum
      "Divide X by Y. `(/ 10.0 2.0)` `(/ 8 3)`"
+  where
+    divide' :: RNativeFun e
+    divide' = binop (\a b -> assert (b /= 0) "Division by 0" $ liftOp (/) a b)
+                    (\a b -> assert (b /= 0) "Division by 0" $ liftOp div a b)
+
+powDef :: NativeDef
 powDef = defRNative "^" pow coerceBinNum "Raise X to Y power. `(^ 2 3)`"
+  where
+    pow :: RNativeFun e
+    pow = binop (\a b -> liftDecF (**) a b)
+                (\a b -> assert (b >= 0) "Integral power must be >= 0" $ liftOp (^) a b)
+
+logDef :: NativeDef
 logDef = defRNative "log" log' coerceBinNum "Log of Y base X. `(log 2 256)`"
+  where
+    log' :: RNativeFun e
+    log' = binop (\a b -> liftDecF logBase a b)
+                 (\a b -> liftIntF logBase a b)
 
+sqrtDef :: NativeDef
 sqrtDef = defRNative "sqrt" (unopd sqrt) unopTy "Square root of X. `(sqrt 25)`"
+
+lnDef :: NativeDef
 lnDef = defRNative "ln" (unopd log) unopTy "Natural log of X. `(round (ln 60) 6)`"
+
+expDef :: NativeDef
 expDef = defRNative "exp" (unopd exp) unopTy "Exp of X `(round (exp 3) 6)`"
+
+absDef :: NativeDef
 absDef = defRNative "abs" abs' (unaryTy tTyDecimal tTyDecimal <> unaryTy tTyInteger tTyInteger)
      "Absolute value of X. `(abs (- 10 23))`"
 
+roundDef :: NativeDef
 roundDef = defTrunc "round" "Performs Banker's rounding" round
+
+ceilDef :: NativeDef
 ceilDef = defTrunc "ceiling" "Rounds up" ceiling
+
+floorDef :: NativeDef
 floorDef = defTrunc "floor" "Rounds down" floor
 
 gtDef, ltDef, gteDef, lteDef, eqDef, neqDef :: NativeDef
```

Comment:
Finish these up for consistency?

Reply:
I just did `modDef`, `absDef`, and `notDef'`. Any others I missed?